Added tests for actionmanagement.c functions

SnowballAntrobus opened the pull request:
Changes:

- Bug fixed action_type_free as it was preventing compilation when used
- Fixed do_item_item_action so that it now also checks to see if the direct item supports the action
- Created a new file test_action_new_init_free.c that includes tests for the new, init, and free action_type functions, using criterion.
- Created a new file test_item_action.c that includes tests for the do_item_action function. They test for correct action kind and allowed_action on the item, using criterion. 
- Created a new file test_item_item_action.c that includes tests for the do_item_item_action function. They test for correct action kind and allowed_action on the items, using criterion. 
- Created a new file test_path_action.c that includes tests for the do_path_action function. They test for correct action kind. 
- altered action functions to return an int descriptive of the status of the action called (#245 ) with a status string (success/failure) as an out parameter
- altered test c files to reflect the altered returned format (#286 , #287 , #288 )
- altered cli code to reflect changes from #245

SnowballAntrobus commented:
Yoon's comments on the do_item_item_action tests PR #204 that apply to other test files:

- malloc size in bytes was replaced with a macro scale variable BUFFER_SIZE in the actions/dev branch. I think you should merge actions/dev into this branch
- all "strcat"s have been replaced with sprintf()
- there will need to be two else if's given the new code in actions/dev, returns 1 if there is one particular error, returns 2 if there is another particular error. We will also need tests to check for failure corresponding to the error codes 1 and 2.
- I just realized that using booleans for return codes may be confusing.
SUCCESS = 0, FAILURE = 1; but true = 1, false = 0. Since we're using the failure return codes 1 and 2, I think you should change rc just to be an integer.

SnowballAntrobus commented:
Made changes in response to Yoon's comments that I listed in my previous comment.

yoonsjeong commented on src/action_management/src/actionmanagement.c:
```diff
@@ -122,8 +120,16 @@ char *do_item_item_action(game_t *g, action_type_t *a,
-    int allowed = allowed_action(indirect, a->c_name);
+    // checks if the action can be used on the direct item
+    int allowed = allowed_action(direct, a->c_name);
+    if (allowed != SUCCESS)
```
it appears there was a duplication error of some sort during the merge! (see line 133)

MTUC08 replied:
This was originally meant to check that the action could be used on both indirect and direct items, but I agree that the direct check should be removed in later PRs since the game state implementation of conditions would have us just check that for example, the item key exists and is associated with the action use, and can be used on the door. For this PR, this is fine.

yoonsjeong commented on src/action_management/src/actionmanagement.c:
```diff
@@ -122,8 +120,16 @@ char *do_item_item_action(game_t *g, action_type_t *a,
+    }
+    // checks if the action can be used on the indirect item
+    allowed = allowed_action(indirect, a->c_name);
     if (allowed != SUCCESS) {
```
if statement here is identical to line 125

MTUC08 replied:
See above

yoonsjeong commented on src/action_management/tests/test_item_actions.c:
```diff
@@ -0,0 +1,101 @@
+    }
+    else if (strcmp(do_item_action(g, a, item), allowed_error) == 0)
+    {
+        rc = 2;
```
might not hurt to define macros for 1, 2, and 4 as well since FAILURE can come in many forms -- I just checked our sprint 4 (not yet on dev) code,
we have
`#define WRONG_KIND 1`
`#define NOT_ALLOWED 2`
I think it wouldn't hurt to add this to the tests
`#define TEST_FAIL 4`
or something similar. is there a reason you used the rc = 4 rather than 3? (3 is currently not being used by any other returns)

SnowballAntrobus replied:
This has been addressed in the new PR #291 . Do you think this can wait? The reason I'm using 4 is because in the tests for item_item_action we use 3 as error code NOT_ALLOWED_INDIRECT.

mpcsh replied:
This definitely can't wait - please make this change as part of this PR.

MTUC08 replied:
@mpcsh Can you elaborate on what change you want us to make? The change to returning different macros is done in #245 and was going to be the PR done after this one. Should we instead add those files to this PR?

mpcsh replied:
Yes - you can do that by just merging that branch into `dev`

MTUC08 replied:
Just created PR #294 for this purpose

ecoble replied:
Would it be possible to just add the macros here and leave the other changes as a separate PR? I wanted to put a CLI reviewer on it and I know the review is mostly done here so I don't want to hold it up.

mpcsh replied:
Looks like @SnowballAntrobus did just that

yoonsjeong commented on src/action_management/tests/test_item_item_actions.c:
```diff
@@ -0,0 +1,131 @@
+    }
+    else if (strcmp(do_item_item_action(g, a, direct, indirect), kind_error) == 0)
+    {
+            rc = FAILURE;
```
I do wonder if we could just have different macros for each test so they don't even depend on common.h
-- so for example at the beginning of each test framework
```
#define TEST_SUCCESS 0
#define KIND_ERROR 1
#define DIRECT_ERROR 2
#define INDIRECT_ERROR 3
#define OUTPUT_ERROR 4
```

yoonsjeong replied:
and then since the test files don't depend on each other (at least, I believe they don't) this shouldn't lead to compilation errors

MTUC08 replied:
At that point, we might as well add a AM common file

yoonsjeong replied:
@SnowballAntrobus thoughts?
I will approve the changes for now since I don't see anything really holding this PR back now

SnowballAntrobus replied:
I think it would be good to approve this for now and make adding a common file an issue. Because as I said in my last comment this was fixed in #291 .

ecoble commented on src/action_management/tests/test_item_actions.c:
```diff
@@ -0,0 +1,101 @@
+#include <criterion/criterion.h>
```
These tests (as well as test_path_actions and test_item_item_actions) feel like they are bordering more on integration tests than unit tests, especially with all the dummy structs and stuff you guys are creating. Not sure if this is a problem, but I know for CLI Borja told us to try and stick to unit testing. Also people are going to start working on a shared dummy game that people can use for testing in the future, so you might want to keep a look out for that or even help build it, and use that for these tests later on instead.

SnowballAntrobus replied:
It was my understanding from talking to Borja and Mark that these were our unit tests. But to be honest, now that you mention it I'm not entirely sure. It's just that the functions won't run without them.

mpcsh replied:
I have kind of the opposite question - I think these are fine in terms of scope (defining test fixtures doesn't make something not a unit test), but I'm just not sure what exactly is being tested here. @SnowballAntrobus can you elaborate on what this is supposed to be checking?

SnowballAntrobus replied:
@mpcsh we are testing our do_action functions sans the conditions and effects from sprint 4. So really we are testing that the correct string is outputted. Like the PR description says, hopefully it’s not terrible, this means we test for validations. This means validating kind and validating that the action is allowed by the item.

mpcsh replied:
Right, ok. Makes sense.

ecoble commented on src/action_management/tests/test_action_new_init_free.c:
```diff
@@ -0,0 +1,153 @@
+    char* c_name = "TAKE";	
+    enum action_kind kind = ITEM;	
+
+     rc = action_type_init(&a, c_name, kind);	
```
Small thing, but this line and the following 2 lines are not properly aligned

ecoble commented on src/action_management/tests/test_action_new_init_free.c:
```diff
@@ -0,0 +1,153 @@
+    char* c_name = "GO";	
+    enum action_kind kind = PATH;	
+
+     rc = action_type_init(&a, c_name, kind);	
```
same here

ecoble commented on src/action_management/tests/test_action_new_init_free.c:
```diff
@@ -0,0 +1,153 @@
+    char* c_name = "PUT_ON";	
+    enum action_kind kind = ITEM_ITEM;	
+
+     rc = action_type_init(&a, c_name, kind);	
```
same here

MTUC08 replied:
note: can be done by running the command astyle --style=stroustrup <filename> in a linux environment

mpcsh commented on src/action_management/tests/test_action_new_init_free.c:
```diff
@@ -0,0 +1,150 @@
+    action_type_free(new_action);
+}
+
+/* Checks that a new action type, kind 1 is initiated without interruption */	
```
`s/initiated/initialized/`
(make sure that gets changed everywhere)

MTUC08 replied:
@mpcsh what exactly does this mean?

lucyyutingli replied:
If I read this correctly it was to change initiated to initialized. I just changed them in the tests.

MTUC08 commented:
Update: I will add a PR to merge the changed return format functions/tests into actions/dev for this PR

SnowballAntrobus commented:
Addressed @mpcsh comments

mpcsh commented:
@SnowballAntrobus one of my change requests wasn't addressed - please make sure that gets taken care of

mpcsh commented:
:tada:

SnowballAntrobus commented:
Closing Statement:
This PR is mostly tests, including new, init, and free tests for the action struct, and testing that the functions for each of the three kinds of actions function properly (do_item_action, do_path_action, and do_item_item_action). We also updated these action functions to return an error code int (0 for success, 1 for incorrect kinds, 2 for invalid direct object and action pairings, 3 for invalid indirect object and action pairings, and 4 for incorrect strings). We updated any files using these functions accordingly, including our tests and CLI files that called these functions. There are also bug fixes for action_type_free and do_item_item_action. In the future we will also update these functions to check for effects and conditions once game state's new functions come out.

mpcsh commented:
Issue Score: ✓+

Comments:
Great work!
<!-- {"score": "check-plus"} -->
